crates/pep508-rs/src/marker/algebra.rs

```diff
@@ -748,6 +760,81 @@ impl Edges {
     }
 }
 
+/// Returns the equivalent `python_full_version` specifier for a `python_version` comparison.
+///
+/// Returns `Err` with a constant node if the equivalent comparison is always `true` or `false`.
+fn python_version_to_full_version(specifier: VersionSpecifier) -> Result<VersionSpecifier, NodeId> {
+    let mut release = specifier.version().release();
+
+    // Skip trailing `0`s after the major and minor segment.
+    if let Some(end) = release.iter().rposition(|segment| *segment != 0) {
+        if end > 0 {
+            release = &release[..=end];
+        }
+    }
+
+    let major_minor = match *release {
+        // Note that `python_version == 3` matches `3.0.1`, `3.0.2`, etc.
+        [major] => Some((major, 0)),
+        [major, minor] => Some((major, minor)),
+        _ => None,
+    };
+
+    if let Some((major, minor)) = major_minor {
+        let version = Version::new([major, minor]);
+
+        Ok(match specifier.operator() {
+            // `python_version == 3.7` is equivalent to `python_full_version == 3.7.*`.
+            Operator::Equal | Operator::ExactEqual => {
+                VersionSpecifier::equals_star_version(version)
+            }
+            // `python_version != 3.7` is equivalent to `python_full_version != 3.7.*`.
+            Operator::NotEqual => VersionSpecifier::not_equals_star_version(version),
+
+            // `python_version > 3.7` is equivalent to `python_full_version >= 3.8`.
+            Operator::GreaterThan => {
+                VersionSpecifier::greater_than_equal_version(Version::new([major, minor + 1]))
+            }
+            // `python_version < 3.7` is equivalent to `python_full_version < 3.7`.
+            Operator::LessThan => specifier,
+            // `python_version >= 3.7` is equivalent to `python_full_version >= 3.7`.
+            Operator::GreaterThanEqual => specifier,
+            // `python_version <= 3.7` is equivalent to `python_full_version < 3.8`.
+            Operator::LessThanEqual => {
+                VersionSpecifier::less_than_version(Version::new([major, minor + 1]))
+            }
+
+            // `==3.7.*`, `!=3.7.*`, `!=3.7` already represent the equivalent `python_full_version`
```

Comment:
Hmm... should `python_version != 3.7` be: `python_version_full <= 3.6 or python_version_full >= 3.8`?

`python_version_full != 3.7` would allow `3.7.1`.

Reply:
Yes, that happens in a branch above; `python_version != 3.7` becomes `python_full_version != 3.7.*`. The comment here is wrong, it applies to `~=` not `!=`. `python_version ~= 3.7` is equivalent to `python_full_version ~= 3.7` I believe.

Reply:
So `python_version ~= 3.7` means `python_version >= 3.7, < 4` -- is that right?

I think you're right then...